app/src/tx_metadata.c

```diff
@@ -0,0 +1,189 @@
+        { \
+            if (!(parsed < txMetadataLength)) { \
+                return PARSER_TEMPLATE_ERROR; \
+            } \
+            (*where) = txMetadata[parsed++]; \
+        } 
+    #define READ_STRING(dest_pointer, len) { \
+        *(len) = 0; \
+        *(dest_pointer) = (char *) txMetadata + parsed; \
+        while (*(len) <= MAX_TEMPLATE_STRING_LENGTH) { \
+            uint8_t byte = 0; \
+            READ_CHAR(&byte); \
+            if (byte == 0) { \
+                break; \
+            } \
+            (*(len))++; \
+        } \
+        if (*(len) > MAX_TEMPLATE_STRING_LENGTH) { \
+            return PARSER_TEMPLATE_ERROR; \
+        } \
+    }
+    #define READ_SKIP(count) { \
+        parsed += (count); \
+    }
+
+    //read number of hashes and validate script
+    {
+        uint8_t numberOfHashes = 0;
+        READ_CHAR(&numberOfHashes)
+        if (numberOfHashes > MAX_TEMPLATE_NUMBER_OF_HASHES) {
+            return PARSER_TEMPLATE_TOO_MANY_HASHES;
+        }
+        parser_error_t err = _validateHash(scriptHash, txMetadata, txMetadataLength);
+        if (err != PARSER_OK) {
+            return err;
+        }
+        READ_SKIP(numberOfHashes*SCRIPT_HASH_SIZE)
+    }
+
+    //read tx name
+    READ_STRING(&parsedTxMetadata->txName, &parsedTxMetadata->txNameLength)
+
+    //read Arguments
+    {
+        READ_CHAR(&parsedTxMetadata->argCount)
+        if (parsedTxMetadata->argCount > PARSER_MAX_ARGCOUNT) {
+            return PARSER_TEMPLATE_TOO_MANY_ARGUMENTS;
+        }
+        STATIC_ASSERT(sizeof(parsedTxMetadata->arguments) >= PARSER_MAX_ARGCOUNT, "Too few arguments in parsed_tx_metadata_t.");
+        for(int i=0; i<parsedTxMetadata->argCount; i++) {
+            READ_CHAR(&parsedTxMetadata->arguments[i].argumentType);
+            READ_STRING(&parsedTxMetadata->arguments[i].displayKey, &parsedTxMetadata->arguments[i].displayKeyLength)
+            READ_CHAR(&parsedTxMetadata->arguments[i].argumentIndex);
+            READ_STRING(&parsedTxMetadata->arguments[i].jsonExpectedType, &parsedTxMetadata->arguments[i].jsonExpectedTypeLength)
+            READ_CHAR(&parsedTxMetadata->arguments[i].jsonExpectedKind);
+        }
+    }
+
+
+    #undef READ_CHAR 
+    #undef READ_STRING
+
+    if (parsed != txMetadataLength) {
+        return PARSER_TEMPLATE_ERROR;
+    }
+
+    return PARSER_OK;
+}
+
+parser_error_t matchStoredTxMetadata(uint8_t scriptHash[SCRIPT_HASH_SIZE], parsed_tx_metadata_t *parsedTxTempate) {
+    size_t i=0;
+    while(KNOWN_TX_METADATA_[i].metadata != NULL) {
+        parser_error_t err = _validateHash(scriptHash, PIC(KNOWN_TX_METADATA_[i].metadata), KNOWN_TX_METADATA_[i].metadataLength);
+        switch (err) {
+            case PARSER_OK:
+                return parseTxMetadata(scriptHash, PIC(KNOWN_TX_METADATA_[i].metadata), KNOWN_TX_METADATA_[i].metadataLength, parsedTxTempate);
+            case PARSER_UNEXPECTED_SCRIPT:
+                break;  //break switch
```

Comment:
Perhaps the comment should instead address why we want to continue the loop even when an error occured.